crates/pathfinder/src/sync/track.rs

```diff
@@ -143,10 +169,99 @@ impl HeaderFanout {
+            transactions: ReceiverStream::new(t_rx).boxed(),
         }
     }
 }
 
+struct TransactionSource {
+    p2p: P2PClient,
+    headers: BoxStream<'static, BlockHeader>,
+    chain_id: ChainId,
+}
+
+impl TransactionSource {
+    fn spawn(self) -> SyncReceiver<BlockTransactions> {
+        let (tx, rx) = tokio::sync::mpsc::channel(1);
+        tokio::spawn(async move {
+            let Self {
+                p2p,
+                mut headers,
+                chain_id,
+            } = self;
+
+            while let Some(header) = headers.next().await {
+                let (peer, mut transactions) = loop {
+                    if let Some(stream) = p2p.clone().transactions_for_block(header.number).await {
+                        break stream;
+                    }
+                };
+
+                let transaction_count = header.transaction_count;
+                let mut transactions_vec = Vec::new();
+
+                // Receive the exact amount of expected events for this block.
+                for _ in 0..transaction_count {
+                    let (transaction, receipt) = match transactions.next().await {
+                        Some(Ok((transaction, receipt))) => (transaction, receipt),
+                        Some(Err(_)) => {
+                            let err = PeerData::new(peer, SyncError2::InvalidDto);
+                            let _ = tx.send(Err(err)).await;
+                            return;
+                        }
+                        None => {
+                            let err = PeerData::new(peer, SyncError2::TooFewTransactions);
+                            let _ = tx.send(Err(err)).await;
+                            return;
+                        }
+                    };
+
+                    transactions_vec.push((transaction, receipt));
+                }
+
+                // Ensure that the stream is exhausted.
+                if transactions.next().await.is_some() {
+                    let err = PeerData::new(peer, SyncError2::TooManyTransactions);
+                    let _ = tx.send(Err(err)).await;
+                    return;
+                }
+
+                let transactions = match compute_hashes(
+                    PeerData {
+                        peer,
+                        data: (header.number, transactions_vec),
+                    },
+                    chain_id,
+                )
+                .await
+                {
+                    Ok(data) => data.data.1,
+                    Err(_) => {
+                        let err = PeerData::new(
+                            peer,
+                            SyncError2::Other(
+                                anyhow!("failed to compute transaction hashes").into(),
+                            ),
+                        );
+                        let _ = tx.send(Err(err)).await;
+                        return;
+                    }
+                };
```

Comment:
This should be moved to a separate stage imo. I know the `checkpoint` sync has this convenient async function for it; but I think we should also be moving it towards a model of stages.

The reason we want this is so that we can decouple receiving data (IO) from the processing stages (hashing in this case), and have buffers in-between. This (hopefully) improves throughput overall.